async_service/asyncio.py

```diff
@@ -32,12 +32,20 @@ class FunctionTask(BaseFunctionTask):
     #
     async def run(self) -> None:
         try:
-            await self._async_fn(*self._async_fn_args)
+            # Name the task, to make it easier to identify which task is holding
+            #   onto the event loop for too long. Also, force an event loop cycle.
+            await asyncio.create_task(
+                self._async_fn(*self._async_fn_args),
+                name=self.name,
+            )
             if self.daemon:
                 raise DaemonTaskExit(f"Daemon task {self} exited")
 
             while self.children:
-                await tuple(self.children)[0].wait_done()
+                await asyncio.gather(*(
+                    child.wait_done()
```

Comment:
I'm not sure this is an improvement.

`self.children` retains all of the active child tasks...  As they each finish they are removed from the list.  In this part of the code what we care about is not blocking the event loop while we wait for all the child tasks to complete.  My intuition says that the `asyncio.gather` approach is going to be heavier while not actually giving us any change in behavior since it will have to orchestrate the machinery of awaiting all of the tasks rather than a single one.

Thoughts?

Reply:
I guess it would require performance testing. If there are 100 children, and they happen to complete in order, then this loop would re-activate 100 times, instead of only reactivating once. Not sure how that compares to whatever overhead `gather()` might add.

Without performance testing, I think the `gather()` approach is easier to read. It's obvious what it's doing: wait for everything to complete. When looking at the current code, I have follow-up questions, like why is the current code looking at only the first one and then looping again? (It requires understanding that the children disappear from the list on their own, which I didn't know/remember).

Reply:
:+1: I'll make the *light* suggestion that we could keep it as it was and augment with a comment to explain why it is implemented that way.  

I'm only -0 on the `asyncio.gather` approach because of some gut-based confidence that it will be less performant in the vast majority of cases since it will create a new task for every child, run it using `ensure_future` which means more event loop context switching.

Reply:
Here's the implementation of `asyncio.gather` btw if you're curious

https://github.com/python/cpython/blob/4b8032e5a4994a7902076efa72fca1e2c85d8b7f/Lib/asyncio/tasks.py#L718